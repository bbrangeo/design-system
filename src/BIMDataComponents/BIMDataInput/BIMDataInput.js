import BIMDataInput from './BIMDataInput.vue';

export default BIMDataInput;