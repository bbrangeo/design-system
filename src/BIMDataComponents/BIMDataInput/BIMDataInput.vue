<template>
  <div
    class="bimdata-input"
    :class="{ error, success, disabled, loading, 'not-empty': !!this.text }"
  >
    <input
      ref="input"
      :id="`bimdata-input-${_uid}`"
      @input="$emit('input', $event.currentTarget.value)"
      :disabled="disabled"
      :value="text"
      @focus="$event.target.select()"
      v-bind="$attrs"
    />
    <div class="bimdata-input__icon">
      <slot name="inputIcon"></slot>
    </div>
    <label :for="`bimdata-input-${_uid}`">{{ placeholder }}</label>
    <span class="bar"></span>
    <span v-if="error" class="error">{{ errorMessage }}</span>
    <span v-if="success" class="success">{{ successMessage }}</span>
  </div>
</template>

<script>
export default {
  model: {
    prop: "text",
    event: "input"
  },
  created() {
    this.$watch(
      () => this.success && this.error,
      successAndError => {
        if (successAndError)
          throw "Input state cannot be both success and error.";
      }
    );
  },
  props: {
    text: {
      type: [String, Number, Boolean]
    },
    placeholder: {
      type: String,
      default: ""
    },
    error: {
      type: Boolean,
      default: false
    },
    success: {
      type: Boolean,
      default: false
    },
    errorMessage: {
      type: String,
      default: ""
    },
    successMessage: {
      type: String,
      default: ""
    },
    loading: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    focus() {
      this.$refs.input && this.$refs.input.focus();
    },
    blur() {
      this.$refs.input && this.$refs.input.blur();
    }
  }
};
</script>

<style lang="scss">
// import BIMDATA STYLE
@import "@/assets/scss/_BIMDataVariables.scss";
@import "@/assets/scss/mixins/_font-size.scss";
@import "@/assets/scss/mixins/_pseudo.scss";

// import COMPONENT STYLE
@import "./_keyframes.scss";
@import "./_BIMDataInput.scss";
</style>
