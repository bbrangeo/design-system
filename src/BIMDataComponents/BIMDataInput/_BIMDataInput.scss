input:-webkit-autofill {
  box-shadow: 0 0 0px 1000px $color-white inset;
}

.bimdata-input {
  height: 32px;
  min-height: 32px;
  margin: calc(#{$spacing-unit} * 2) 0;
  position: relative;
  &.not-empty {
    label {
      top: -22px;
      color: $color-primary;
      @include font-size(10px);
    }
  }
  input {
    padding: 0;
    width: 100%;
    height: 100%;
    border: none;
    display: flex;
    cursor: pointer;
    @include font-size(14px);
    &:focus {
      outline: none;
      & ~ label {
        top: -22px;
        @include font-size(10px);
      }
      & ~ .bar {
        &:after {
          width: 100%;
          transition: transform 0.2s cubic-bezier(0.55, 0, 0.55, 0.2);
          transform: scaleX(1);
        }
      }
    }
  }
  &__icon{
    position: absolute;
    right: 0;
    top: 8px;
  }
  label {
    position: absolute;
    top: 0;
    left: 0;
    @include font-size(12px);
    transform: translateY(8px);
    transition: 0.2s ease all;
    color: $color-tertiary-dark;
    cursor: pointer;
  }
  .bar {
    width: 100%;
    height: 1px;
    position: absolute;
    display: block;
    background: $color-tertiary;
    bottom: 0;
    &::after, &::before {
      width: 0;
      height: 2px;
      @include pseudo;
      bottom: 0;
      left: 0;
      right: 0;
      background-color: $color-primary;
      transform: scaleX(0);
    }
  }
  &.loading input ~ .bar {
    &:after {
      width: 50%;
      animation: scaleX 2s linear infinite none;
      transform-origin: right;
    }
    &:before {
      width: 50%;
      animation: scaleX 2s linear 1s infinite none;
      transform-origin: right;
    }
  }
  .error {
    top: 2px;
    position: relative;
    color: $color-high;
    @include font-size(11px);
  }
  .success {
    top: 2px;
    position: relative;
    color: $color-success;
    @include font-size(11px);
  }
  &.disabled {
    opacity: 0.6;
  }
  &.error {
    label {
      color: $color-high;
    }
    .bar {
      background-color: $color-high;
      &::before,
      &::after {
        background-color: $color-high;
      }
    }
  }
  &.success {
    label {
      color: $color-success;
    }
    .bar {
      background-color: $color-success;
      &::before,
      &::after {
        background-color: $color-success;
      }
    }
  }
}
