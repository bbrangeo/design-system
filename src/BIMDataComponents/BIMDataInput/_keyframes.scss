@keyframes scaleX {
  0% {
    transform: scaleX(0);
    left: -50%;
  }
  25% {
    transform: scaleX(1);
    left: 0;
  }
  50% {
    transform: scaleX(1);
    left: 50%;
  }
  75% {
    transform: scaleX(0);
    left: 50%;
  }
  100% {
    transform: scaleX(0);
    left: 50%;
  }
}
