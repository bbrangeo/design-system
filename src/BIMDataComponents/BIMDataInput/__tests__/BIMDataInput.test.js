import { shallowMount } from "@vue/test-utils";
import { mount } from "@vue/test-utils";
import BIMDataInput from "../BIMDataInput.vue";

describe("BIMDataInput", () => {
  it("should render component and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render component with placeholder and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput, {
      propsData: {
        placeholder: "input placeholder",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render component with error message and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput, {
      propsData: {
        error: true,
        errorMessage: "error message",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render component with success message and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput, {
      propsData: {
        success: true,
        successMessage: "success message",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render loading component and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput, {
      propsData: {
        loading: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render disabled component and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput, {
      propsData: {
        disabled: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render component with 'not empty' class and match snapshot", () => {
    const wrapper = shallowMount(BIMDataInput, {
      propsData: {
        text: "text message",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should work with v-model binding", async () => {
    const onInput = jest.fn();
    const wrapper = shallowMount(BIMDataInput, {
      listeners: {
        input: onInput,
      },
    });
    expect(wrapper.find("input").element.value).toBe("");

    wrapper.find("input").setValue("x");
    expect(onInput).toHaveBeenCalledWith("x");

    wrapper.setProps({ text: "message" });
    await wrapper.vm.$nextTick();
    expect(wrapper.find("input").element.value).toBe("message");
  });

  test("should render a button with a slot and match snapshot", () => {
    const wrapper = mount(BIMDataInput, {
      slots: {
        inputIcon: "test icon slot",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
