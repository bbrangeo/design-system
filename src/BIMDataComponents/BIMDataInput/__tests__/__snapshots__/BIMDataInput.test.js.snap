// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BIMDataInput should render a button with a slot and match snapshot 1`] = `
"<div class=\\"bimdata-input\\"><input id=\\"bimdata-input-17\\">
  <div class=\\"bimdata-input__icon\\">test icon slot</div> <label for=\\"bimdata-input-17\\"></label> <span class=\\"bar\\"></span>
  <!---->
  <!---->
</div>"
`;

exports[`BIMDataInput should render component and match snapshot 1`] = `
"<div class=\\"bimdata-input\\"><input id=\\"bimdata-input-1\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-1\\"></label> <span class=\\"bar\\"></span>
  <!---->
  <!---->
</div>"
`;

exports[`BIMDataInput should render component with 'not empty' class and match snapshot 1`] = `
"<div class=\\"bimdata-input not-empty\\"><input id=\\"bimdata-input-13\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-13\\"></label> <span class=\\"bar\\"></span>
  <!---->
  <!---->
</div>"
`;

exports[`BIMDataInput should render component with error message and match snapshot 1`] = `
"<div class=\\"bimdata-input error\\"><input id=\\"bimdata-input-5\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-5\\"></label> <span class=\\"bar\\"></span> <span class=\\"error\\">error message</span>
  <!---->
</div>"
`;

exports[`BIMDataInput should render component with placeholder and match snapshot 1`] = `
"<div class=\\"bimdata-input\\"><input id=\\"bimdata-input-3\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-3\\">input placeholder</label> <span class=\\"bar\\"></span>
  <!---->
  <!---->
</div>"
`;

exports[`BIMDataInput should render component with success message and match snapshot 1`] = `
"<div class=\\"bimdata-input success\\"><input id=\\"bimdata-input-7\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-7\\"></label> <span class=\\"bar\\"></span>
  <!----> <span class=\\"success\\">success message</span>
</div>"
`;

exports[`BIMDataInput should render disabled component and match snapshot 1`] = `
"<div class=\\"bimdata-input disabled\\"><input id=\\"bimdata-input-11\\" disabled=\\"disabled\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-11\\"></label> <span class=\\"bar\\"></span>
  <!---->
  <!---->
</div>"
`;

exports[`BIMDataInput should render loading component and match snapshot 1`] = `
"<div class=\\"bimdata-input loading\\"><input id=\\"bimdata-input-9\\">
  <div class=\\"bimdata-input__icon\\"></div> <label for=\\"bimdata-input-9\\"></label> <span class=\\"bar\\"></span>
  <!---->
  <!---->
</div>"
`;
